### PR TITLE
[Issue #21] Adds the ability to put count and ratio in the layout.

### PR DESCRIPTION
**Pull request by hfytr**
This PR adds a --layout-format flag which enables setting the layout to an arbitrary string containing count {c}, ratio {r}, and layout image {l}. for example passing --layout-format "{c} {r}\n{l}" may result in something like:
```
3 0.5
│├──┤
```
This flag is optional, and takes the default value {l}, which should have the same behaviour as before this PR. Within the string, brackets are allowed using \\{, \\}. \n, \r, \t, \v, and nerdfont icons are also supported.

Tests have not been written, and I am definitely not a C programmer, so I would be more surprised than not if the code has no vulnerabilities / edge cases that I am missing.

NOTE 1: I committed the flake.lock / flake.nix for future contributors, but its up to you if you want to keep them.

NOTE 2: There is no clang-format, so my formatting is likely wrong.

### CODE REVIEW

**Comment by alex-courtis** [comment]
This looks great - very easy to define.

I'll test and review tomorrow.

**Comment by alex-courtis** [comment]
Happy to help out with tests; they should be straightforward unit.

**Comment by hfytr** [comment]
Hello, it seems that I ran make test, and then make iwyu. Incorporating the suggestions of make iwyu broke the tests. Anyway, should be good now.

**Comment by alex-courtis** [comment]
> Hello, it seems that I ran make test, and then make iwyu. (thoughtlessly) Incorporating the suggestions of make iwyu broke the tests. Anyway, should be good now.

Inconsistent iwyu is becoming a problem, I think I'll just remove it.

I've just made a config change #23 to cppcheck, for versions greater than 2.7 (CI)

If it's OK with you I'll:
* update this branch
* write and push tests

**Comment by hfytr** [comment]
Looks good! Thanks for putting up with me not following CONTRIBUTING.md. For future, how do I run checks locally?

**Comment by alex-courtis** [comment]
> Looks good! Thanks for putting up with me not following CONTRIBUTING.md. For future, how do I run checks locally?

You're all good. It looks like you've been running all of them. `cppcheck` is the useful one: https://github.com/alex-courtis/wideriver/blob/master/CONTRIBUTING.md#lint

Not sure how your environment's setup, but you can manually check cross-compiler:
```sh
make CC=gcc clean all
make CC=clang clean all
```

I'll deal with iwyu

File: src/layout.c [hunk 1]
```diff
@@ -54,6 +56,88 @@ const char *description_info(const struct Demand* const demand, const struct Tag
...
+
+		escaped -= 1;
+		if (escaped < 0)
+			escaped = 0;
+	}
+
+	return desc;
+}
+
 const char *description_debug(const struct Demand* const demand, const struct Tag* const tag) {
```
Comment (alex-courtis) [inline src/layout.c:141]: We can delete this now :)

File: inc/enum.h [hunk 2]
```diff
@@ -27,6 +27,14 @@ enum LogThreshold {
 	LOG_THRESHOLD_DEFAULT = INFO,
 };
 
+enum LayoutFormatElem {
+	LAYOUT = 'l',
+	COUNT = 'c',
+	RATIO = 'r',
```
Comment (alex-courtis) [inline inc/enum.h:33]: Can we please add `{n}` for name i.e. wide, left etc. ?

This will allow users to extract the current layout, as the image is variable.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looking good, I'm using it right now.

I've added a test suite for layout and added one basic test for `cfg_set_layout_format`.

Please

- [x] add `{n}`
- [x] delete `description_debug` and move `description_info` into `layout_description`
- [ ] complete test cases for `cfg_set_layout_format`

I'll think about how we can simplify `description_info`. I've seen something similar but I can't recall it right now.

I'm unavailable until next weekend.

**Comment by alex-courtis** [comment]
> NOTE 2: There is no clang-format, so my formatting is likely wrong.

Sorry, I have to get to that...

What IDE do you use? vim `gg=G` is fine.

**Comment by alex-courtis** [comment]
We could use a simple string replacer:

```c
	static char desc[LAYOUT_FORMAT_LEN+5];

	snprintf(sratio, 10, "%g", ratio);
	snprintf(scount, 10, "%d", count);

	char *counted = replace_string(cfg->layout_format, "{c}", scount);

	char *ratioed = replace_string(counted, "{r}", sratio);
	free(counted);

	char *imaged = replace_string(ratioed, "{l}", image);
	free(ratioed);

	strncpy(desc, imaged, LAYOUT_FORMAT_LEN + 5);
	free(imaged);

	return desc;
```

```c
// return a copy of str with instances of target substituted with replacement
// call frees
char *replace_string(const char * const str, const char * const target, const char * const replacement) {
	char *match = strstr(str, target);
	if (!match) {
		return strdup(str);
	}

	// allocate result
	size_t len = strlen(str) + strlen(replacement) - strlen(target) + 1;
	char *res = calloc(len, sizeof(char));

	// find string up to match
	size_t len_pre = match - str + 1;
	char pre[len_pre];
	snprintf(pre, len_pre, "%s", str);

	// format
	snprintf(res, len, "%s%s%s", pre, replacement, match + strlen(target));

	// recurse
	char *next = replace_string(res, target, replacement);
	free(res);
	return next;
}
```

**Comment by hfytr** [comment]
OK, so I added {n}, changed some things to use strn functions. my impl of {n} is pretty bad, as I wasn't sure how to integrate with the existing switch / case. Also, I/you will have to change it to use your string replacer (which LGTM).

**Comment by alex-courtis** [comment]
> OK, so I added {n}, changed some things to use strn functions. my impl of {n} is pretty bad, as I wasn't sure how to integrate with the existing switch / case. Also, I/you will have to change it to use your string replacer (which LGTM).

Looks good, moved it to the very hidden function from enum.c

> EDIT: Your tests were using (the now nonexistent) description_info, so i switched them to layout_description() instead.

Updated layout_description to use string replace. Changed the layout functions to be caller frees, as per the original comments.

> Sorry for the horrible git history.

Don't worry about git history, it will be squashed on merge.

We do need to take care of the documentation - man and readme. I'd be grateful if you could update `doc/templ/40.man.readme.md` with the new options.
Don't worry if you can't run `make doc`, it's pretty hacky.

**Comment by alex-courtis** [comment]
Unfortunately I'm out of time; I'll get back to this next weekend.

This one will be a slow burn...

File: src/usage.c [hunk 3]
```diff
@@ -17,59 +17,77 @@ void usage(const int status) {
 			"\n"
 			"  --stack                         %s|%s|%s                 %s\n"
 			"\n"
-			"  --count-master                  count                                 %d           %d <= count\n"
-			"  --ratio-master                  ratio                                 %.02f      %.1g <= ratio <= %.1g\n"
+			"  --count-master                  count                             "
```
Comment (alex-courtis) [inline src/usage.c:20]: Please don't change formatting for existing lines.

File: src/cfg.c [hunk 4]
```diff
@@ -190,6 +191,39 @@ bool cfg_set_border_width_monocle(const char *s) {
 	}
 }
 
+bool cfg_set_layout_format(const char *s) {
```
Comment (alex-courtis) [inline src/cfg.c:194]: I don't quite understand the failure cases of this function; I might be missing the point, but could it just be "garbage in, garbage out"?

Can you please add tests to `tst-args_cli.c` to cover these cases?

**Comment by nirodhvana** [comment]
@alex-courtis will this be merged?

**Comment by alex-courtis** [comment]
> @alex-courtis will this be merged?

I'm not quite sure of the status of this PR. There is still a lot of unnecessary code such as
* all the `\t` `\r` escaping
* `--layout-format` parsing
  * does not seem to be necessary at all
  * doesn't seem safe from a memory or security perspective

I attempted a rebase and a merge from upstream - ~1 year ago - however I gave up.

The best course of action might be to create a fresh branch and pull in only the minimum required.

As usual, pull requests are most gratefully appreciated, see [CONTRIBUTING.md](https://github.com/alex-courtis/wideriver/blob/master/CONTRIBUTING.md)
